### PR TITLE
[shopsys] new class for representing monetary values PHASE 4

### PR DESCRIPTION
**Pull request by PetrHeinz**
In this PR, we're changing mainly the `Price` class and support the `Money` class in Twig filters.

| Q             | A
| ------------- | ---
|Description, reason for the PR| PR for easier review and management of #821
|New feature| Yes 
|[BC breaks](/docs/contributing/backward-compatibility-promise.md)| Yes 
|Fixes issues| see #821 
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/tests/Test/IsMoneyEqual.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Tests\FrameworkBundle\Test;
+
+use PHPUnit\Framework\Constraint\Constraint;
+use Shopsys\FrameworkBundle\Component\Money\Money;
+
+class IsMoneyEqual extends Constraint
```
Comment (simara-svatopluk) [inline packages/framework/tests/Test/IsMoneyEqual.php]: There are missed possible useses:

`BasePriceCalculationTest, PaymentPriceCalculationTest, TransportPriceCalculationTest`

you can find them by regexp `.*True.*get.*->equa`

File: project-base/src/Shopsys/ShopBundle/Resources/views/Front/Content/Order/preview.html.twig [hunk 2]
```diff
@@ -20,7 +20,7 @@
                         {{ 'Promo code'|trans }} {{ (-orderPreview.promoCodeDiscountPercent)|formatPercent }}
                     </td>
                     <td class="table-cart-preview__cell table-cart-preview__cell--price">
-                        {{ (-quantifiedProductDiscount.priceWithVat)|price }}
+                        {{ (-quantifiedProductDiscount.priceWithVat.toValue)|price }}
```
Comment (simara-svatopluk) [inline project-base/src/Shopsys/ShopBundle/Resources/views/Front/Content/Order/preview.html.twig]:
```suggestion
                        {{ quantifiedProductDiscount.priceWithVat.multiply('-1')|price }}
```

I don't know if it works, but it would be more Money-friendly; also right now we are using deprecated behavior

Also I would like to have `negative` method instead of `multiply(-1)` as mentioned earlier :-)

File: packages/framework/src/Resources/views/Admin/Content/Order/listGrid.html.twig [hunk 3]
```diff
@@ -14,7 +14,7 @@
 {% endblock %}
 
 {% block grid_value_cell_id_total_price %}
-    {{ value|priceWithCurrency(row.order.currency) }}
+    {{ value.toValue|priceWithCurrency(row.order.currency) }}
```
Comment (simara-svatopluk) [inline packages/framework/src/Resources/views/Admin/Content/Order/listGrid.html.twig]:
```suggestion
    {{ value|priceWithCurrency(row.order.currency) }}
```

File: packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig [hunk 4]
```diff
@@ -353,6 +353,14 @@
     {{ value|formatDateTime }}
 {% endblock %}
 
+{% block grid_value_cell_type_DateTimeImmutable %}
+    {{ value|formatDateTime }}
+{% endblock %}
+
+{% block grid_value_cell_type_Shopsys_FrameworkBundle_Component_Money_Money %}
+    {{ value.toValue|priceWithCurrencyAdmin }}
```
Comment (simara-svatopluk) [inline packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig]:
```suggestion
    {{ value|priceWithCurrencyAdmin }}
```

File: packages/framework/src/Component/Doctrine/GroupedScalarHydrator.php [hunk 5]
```diff
@@ -51,6 +51,8 @@ protected function gatherGroupedScalarRowData(&$data)
             $fieldName = $cacheKeyInfo['fieldName'];
 
             if (isset($cacheKeyInfo['isScalar'])) {
+                $type  = $cacheKeyInfo['type'];
+                $value = $type->convertToPHPValue($value, $this->_platform);
```
Comment (simara-svatopluk) [inline packages/framework/src/Component/Doctrine/GroupedScalarHydrator.php]: What kind of sorcery is this???

Couple of lines later, there is ` $value = $type ? $type->convertToPHPValue($value, $this->_platform) : $value;` Why don't we check what the `$type` is?

File: packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig [hunk 6]
```diff
@@ -353,6 +353,10 @@
     {{ value|formatDateTime }}
 {% endblock %}
 
+{% block grid_value_cell_type_DateTimeImmutable %}
```
Comment (simara-svatopluk) [inline packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig]: Beware, commit 037b908280312bd62affc62d58abbe982db8c842 is fixuped to wrong commit and should be standalone commit anyway

File: packages/framework/src/Model/Order/Order.php [hunk 7]
```diff
@@ -1165,8 +1165,8 @@ protected function addOrderItemDiscount(
             $orderItem->getOrder(),
             $name,
             new Price(
-                -$quantifiedItemDiscount->getPriceWithoutVat(),
-                -$quantifiedItemDiscount->getPriceWithVat()
+                $quantifiedItemDiscount->getPriceWithoutVat()->multiply('-1'),
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Order/Order.php]: @PetrHeinz is working on this

File: packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php [hunk 8]
```diff
@@ -148,8 +149,8 @@ public function arePricesDifferent(array $prices)
         $firstPrice = array_pop($prices);
         /* @var $firstPrice \Shopsys\FrameworkBundle\Model\Pricing\Price */
         foreach ($prices as $price) {
-            if ($price->getPriceWithoutVat() !== $firstPrice->getPriceWithoutVat()
-                || $price->getPriceWithVat() !== $firstPrice->getPriceWithVat()
+            if (!$price->getPriceWithoutVat()->equals($firstPrice->getPriceWithoutVat())
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php]: Do you think it is necessary? Previous usage was unnecessarily complex and current solution is much clearer.

File: packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php [hunk 9]
```diff
@@ -127,7 +127,8 @@ public function getMinimumPriceByPriceWithoutVat(array $prices)
 
         $minimumPrice = null;
         foreach ($prices as $price) {
-            if ($minimumPrice === null || $price->getPriceWithoutVat() < $minimumPrice->getPriceWithoutVat()) {
+            /** @var \Shopsys\FrameworkBundle\Model\Pricing\Price|null $minimumPrice */
+            if ($minimumPrice === null || $price->getPriceWithoutVat()->isLessThan($minimumPrice->getPriceWithoutVat())) {
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php]: I think that is some nice to have solution and is out of scope of this user story

File: packages/framework/src/Model/Order/Order.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (PetrHeinz) [reply packages/framework/src/Model/Order/Order.php]: Good idea with the method, implemented in 6a9b1dbd3

File: packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (PetrHeinz) [reply packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php]: It's not as easy: Is a price of 120,- with 20% VAT greater or less than 110,- with 0% VAT?

120,- > 110,- but 100,- < 110,-

File: packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig (hunk 4 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig]: Fixed

File: packages/framework/src/Resources/views/Admin/Content/Order/listGrid.html.twig (hunk 3 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/views/Admin/Content/Order/listGrid.html.twig]: Fixed

File: project-base/src/Shopsys/ShopBundle/Resources/views/Front/Content/Order/preview.html.twig (hunk 2 above)
Comment (PetrHeinz) [reply project-base/src/Shopsys/ShopBundle/Resources/views/Front/Content/Order/preview.html.twig]: I've implemented `inverse` to handle this in 6a9b1dbd3

File: packages/framework/src/Component/Doctrine/GroupedScalarHydrator.php (hunk 5 above)
Comment (TomasLudvik) [reply packages/framework/src/Component/Doctrine/GroupedScalarHydrator.php]: This code has been copy pasted from \Doctrine\ORM\Internal\Hydration\AbstractHydrator::gatherRowData:288-289

File: packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Model/Product/Pricing/ProductPriceCalculation.php]: After explaining problems with VATs, I understnd it not the easiest task

File: packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig (hunk 6 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/views/Admin/Grid/Grid.html.twig]: After discussion with @PetrHeinz we well squash ed8714c6a9cc884a7c6a1543e8c0b504b0c2a4ba e2a985401113feb2a0de00e1571c4c89fa446e68 037b908280312bd62affc62d58abbe982db8c842 442c2b570770660c8c607bc550ad9b8bcbe0a81b 02a6f051f22c5ef52808811829864233563b847b 04fb9706006de0bb2394e2d23aa31e6b767a7427

In one commit

File: packages/framework/tests/Test/IsMoneyEqual.php (hunk 1 above)
Comment (TomasLudvik) [reply packages/framework/tests/Test/IsMoneyEqual.php]: Fixed in 3de2a6564737f1fe4dfbdb0a289c600dcd22e565

File: packages/framework/src/Component/Grid/MoneyConvertingDataSourceDecorator.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\FrameworkBundle\Component\Grid;
+
+use Shopsys\FrameworkBundle\Component\Money\Money;
+use Shopsys\FrameworkBundle\Component\Paginator\PaginationResult;
+
+class MoneyConvertingDataSourceDecorator implements DataSourceInterface
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Component/Grid/MoneyConvertingDataSourceDecorator.php]: Well it is decorating the `DataSource` though I think this is best possible name.

File: packages/framework/src/Model/Pricing/Price.php [hunk 11]
```diff
@@ -2,54 +2,64 @@
...
     {
         return $this->priceWithVat;
     }
 
     /**
-     * @return string
+     * @return \Shopsys\FrameworkBundle\Component\Money\Money
      */
-    public function getVatAmount()
+    public function getVatAmount(): Money
```
[the remark this answers is not in the document]
Comment (PetrHeinz) [reply packages/framework/src/Model/Pricing/Price.php]: It's used in Twig, see Front/Content/Order/preview.html.twig:

```twig
<tr class="table-cart-preview__row table-cart-preview__row--total">
    <td class="table-cart-preview__cell">
        {{ 'Total amount of VAT'|trans }}:
    </td>
    <td class="table-cart-preview__cell table-cart-preview__cell--price">
        {{ orderPreview.totalPrice.vatAmount|price }}
    </td>
</tr>
```

File: packages/framework/src/Model/Order/OrderPriceCalculation.php [hunk 12]
```diff
@@ -72,12 +73,12 @@ public function calculateOrderRoundingPrice(
         }
 
         $roundingPrice = $this->rounding->roundPriceWithVat(
-            round($orderTotalPrice->getPriceWithVat()) - $orderTotalPrice->getPriceWithVat()
+            $orderTotalPrice->getPriceWithVat()->round(0)->subtract($orderTotalPrice->getPriceWithVat())->toValue()
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Model/Order/OrderPriceCalculation.php]: We have extracted parts into separate variables so it is easily readable, more info: 9dfc98341bf4dd9a243eef2719623cfc852bd47b

File: packages/framework/src/Controller/Admin/CustomerController.php [hunk 13]
```diff
@@ -186,6 +188,7 @@ public function listAction(Request $request)
         );
 
         $dataSource = new QueryBuilderDataSource($queryBuilder, 'u.id');
+        $dataSource = new MoneyConvertingDataSourceDecorator($dataSource, ['ordersSumPrice']);
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Controller/Admin/CustomerController.php]: We have renamed the first one to $innerDataSource

File: packages/framework/src/Component/Grid/MoneyConvertingDataSourceDecorator.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Component/Grid/MoneyConvertingDataSourceDecorator.php]: We have agreed on keeping this with @PetrHeinz

[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Component/Grid/MoneyConvertingDataSourceDecorator.php]: Even it is decorating, the decoration is just a detail, the main point is that it is "converting money data source". What is added value of the name `Decorator`?

If I take Your logic to extreme, all classes should be called with postfix `Class`, eg. `MoneyConvertingDataSourceDecoratorClass` - it doesn't make sense.

File: packages/framework/src/Model/Order/OrderPriceCalculation.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Model/Order/OrderPriceCalculation.php]: It is still suboptimal solution for me but... egal :-)

File: packages/framework/src/Component/Doctrine/GroupedScalarHydrator.php (hunk 5 above)
Comment (simara-svatopluk) [reply packages/framework/src/Component/Doctrine/GroupedScalarHydrator.php]: Wow, I still don't understand it.... So if it works... why not :-)

File: packages/framework/src/Twig/MoneyExtension.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Twig;
+
+use Shopsys\FrameworkBundle\Component\Money\Money;
+use Twig_Extension;
+use Twig_SimpleFilter;
+
+class MoneyExtension extends Twig_Extension
```
Comment (simara-svatopluk) [inline packages/framework/src/Twig/MoneyExtension.php]: Shouldn't be the functionality part of `PriceExtension`? Because the `PriceExtension` works with `Money` too...

Also I'm not sure about the filter name `moneyFormat` as we use everywhere in the system `price....` filters for `Money`

File: packages/framework/src/Twig/PriceExtension.php [hunk 15]
```diff
@@ -110,7 +113,7 @@ public function getFilters()
     /**
      * @return array
      */
-    public function getFunctions()
+    public function getFunctions(): array
```
Comment (simara-svatopluk) [inline packages/framework/src/Twig/PriceExtension.php]: please add return typehint also to `getFilters()`

Comment (simara-svatopluk) [reply packages/framework/src/Twig/PriceExtension.php]: Awesome commit :tada: 8d4a56ae4f

File: packages/framework/src/Twig/PriceExtension.php [hunk 16]
```diff
@@ -220,16 +235,28 @@ public function priceWithCurrencyByCurrencyIdFilter($price, $currencyId)
...
      * @return string
      */
     private function formatCurrency($price, Currency $currency, $locale = null)
     {
-        if (!is_numeric($price)) {
-            return $price;
+        if ($price instanceof Money) {
+            $price = $price->toString();
+        } else {
+            $message = sprintf('Twig filters for prices must receive a value of "%s" type.', Money::class);
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Twig/PriceExtension.php]: We have removed it

File: packages/framework/src/Twig/MoneyExtension.php (hunk 14 above)
Comment (simara-svatopluk) [reply packages/framework/src/Twig/MoneyExtension.php]: please just document the new `moneyFormat` filter

File: packages/framework/src/Twig/PriceExtension.php (hunk 15 above)
Comment (TomasLudvik) [reply packages/framework/src/Twig/PriceExtension.php]: Fixed

File: packages/framework/src/Twig/MoneyExtension.php [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+     * @param string $decimalPoint
+     * @param string $thousandsSeparator
+     * @return string
+     */
+    public function moneyFormatFilter(Money $money, int $decimal = null, string $decimalPoint = '.', string $thousandsSeparator = '')
+    {
+        $moneyString = $money->toString();
+
+        if ($decimal === null) {
+            $decimal = $this->getNumberOfDecimalPlaces($moneyString);
```
[the remark this answers is not in the document]
Comment (PetrHeinz) [reply packages/framework/src/Twig/MoneyExtension.php]: The usage `$decimal = strlen($money->centsOnly()->toValue()) - 2;` would be a bit cumbersome, because you would have to check `->isZero()` as well, otherwise you could get -1 (`strlen('0') - 2`).

We consider the method as not so important as to include it now.

File: packages/framework/src/Twig/MoneyExtension.php (hunk 14 above)
Comment (PetrHeinz) [reply packages/framework/src/Twig/MoneyExtension.php]: We'll document it in phase 6 :)
